File: Wheel_Maker/Swerve_Evaluate_Case.m
function out = Swerve_Evaluate_Case(caseIn, p)
% SWERVE_EVALUATE_CASE 舵轮底盘单工况全链路评估器
%
% 描述:
%   该函数是底层的最顶层封装。它接收一个明确的运动指令(期望速度+期望加速度)，
%   先调用运动学解算器计算目标舵角和轮速，然后调用动力学引擎核算扭矩和打滑风险，
%   最后将所有结果和中间调试变量打包为一个完整的结构体输出。
%
% 输入:
%   caseIn.vx            - 期望前向速度 (m/s)
%   caseIn.vy            - 期望右向速度 (m/s)
%   caseIn.omega_z       - 期望逆时针角速度 (rad/s)
%   caseIn.ax            - 期望前向真实加速度 (m/s²)
%   caseIn.ay            - 期望右向真实加速度 (m/s²)
%   caseIn.dt            - 控制/评估周期 (s)
%   caseIn.current_theta - 当前四个舵角反馈值 (1x4, rad)
%   caseIn.alpha_steer   - (可选) 强行指定的转向角加速度 (1x4, rad/s²)
%
% 输出:
%   out - 包含所有评估结论 (速度分配、扭矩核算、形变参数、打滑标志) 的综合字典

    if nargin < 2 || isempty(p)
        p = Config_Params.toStruct();
    else
        p = Config_Params.mergeWithDefaults(p);
    end

    % 自动补全缺失的工况输入参数，防止报错
    caseIn = completeCaseDefaults(caseIn, p);
    limits = Swerve_Performance_Limits(p);

    % --- 阶段一：调用运动学内核 (解算速度与角度) ---
    [vDrive, thetaSteer, omegaSteer, kinDbg] = ...
        Swerve_Kinematics_Solver( ...
            caseIn.vx, caseIn.vy, caseIn.omega_z, ...
            caseIn.current_theta, caseIn.dt, p);

    % 确定转向角加速度：若未指定，则通过微分计算，并施加物理饱和限制
    wheel_count = 4;
    if isfield(p, 'swerve_wheel_count')
        wheel_count = p.swerve_wheel_count;
    end

    if isfield(caseIn, 'alpha_steer') && ~isempty(caseIn.alpha_steer)
        alphaSteer = reshape(caseIn.alpha_steer, 1, wheel_count);
    else
        alphaSteer = abs(omegaSteer) ./ max(caseIn.dt, eps);
        alphaLimit = p.swerve_target_max_alpha;
        alphaSteer = min(alphaSteer, alphaLimit);
    end

    % --- 阶段二：调用动力学内核 (解算扭矩与受力形变) ---
    [driveT, steerT, slipFlag, dynDbg] = ...
        Swerve_Dynamics_Core(caseIn.ax, caseIn.ay, alphaSteer, p);

    % --- 阶段三：结果封装聚合 ---
    out = struct();
    out.case = caseIn;
    out.p = p;
    out.limits = limits;

    out.v_drive = vDrive;
    out.theta_steer = thetaSteer;
    out.omega_steer = omegaSteer;
    out.alpha_steer = alphaSteer;

    out.drive_torque = driveT;
    out.steer_torque = steerT;
    out.slip_flag = slipFlag;

    out.kin_debug = kinDbg;
    out.dyn_debug = dynDbg;

    % 提取极值，用于快速状态判断
    out.max_drive_torque = max(driveT);
    out.max_steer_torque = max(steerT);
    out.max_grip_usage = max(dynDbg.grip_usage);
    out.max_tire_delta = max(dynDbg.delta);
    out.max_contact_half_len = max(dynDbg.contact_half_len);

    [~, out.max_grip_wheel_idx] = max(dynDbg.grip_usage);
    [~, out.max_delta_wheel_idx] = max(dynDbg.delta);
    if wheel_count == 3
        out.wheel_names = ["F", "RR", "RL"];
    else
        out.wheel_names = ["FL", "FR", "RR", "RL"];
    end
end

% ---------------- 辅助函数区 ----------------
function caseOut = completeCaseDefaults(caseIn, p)
    if nargin < 1 || isempty(caseIn), caseIn = struct(); end
    caseOut = caseIn;
    caseOut = setDefault(caseOut, 'vx', 0.0);
    caseOut = setDefault(caseOut, 'vy', 0.0);
    caseOut = setDefault(caseOut, 'omega_z', 0.0);
    caseOut = setDefault(caseOut, 'ax', 0.0);
    caseOut = setDefault(caseOut, 'ay', 0.0);
    caseOut = setDefault(caseOut, 'dt', 0.02);

    wheel_count = 4;
    if nargin >= 2 && isfield(p, 'swerve_wheel_count')
        wheel_count = p.swerve_wheel_count;
    end

    caseOut = setDefault(caseOut, 'current_theta', zeros(1, wheel_count));

    caseOut.current_theta = reshape(caseOut.current_theta, 1, wheel_count);
    if ~isfinite(caseOut.dt) || caseOut.dt <= 0, caseOut.dt = 0.02; end
end

function s = setDefault(s, fieldName, defaultValue)
    if ~isfield(s, fieldName) || isempty(s.(fieldName))
        s.(fieldName) = defaultValue;
    end
end

File: web/src/core/Swerve_Evaluate_Case.js
import { mergeWithDefaults } from './Config_Params';
import { Swerve_Performance_Limits } from './Swerve_Performance_Limits';
import { Swerve_Kinematics_Solver } from './Swerve_Kinematics_Solver';
import { Swerve_Dynamics_Core } from './Swerve_Dynamics_Core';

export function Swerve_Evaluate_Case(caseIn = {}, p_in) {
    const p = mergeWithDefaults(p_in);
    const wheel_count = p.swerve_wheel_count || 4;
    
    const vx = caseIn.vx || 0.0;
    const vy = caseIn.vy || 0.0;
    const omega_z = caseIn.omega_z || 0.0;
    const ax = caseIn.ax || 0.0;
    const ay = caseIn.ay || 0.0;
    const dt = (caseIn.dt && caseIn.dt > 0) ? caseIn.dt : 0.02;
    const current_theta = caseIn.current_theta && caseIn.current_theta.length === wheel_count ? caseIn.current_theta : new Array(wheel_count).fill(0);
    
    const limits = Swerve_Performance_Limits(p);
    
    const { v_drive_out, theta_steer_out, omega_steer_out, debug: kin_debug } = 
        Swerve_Kinematics_Solver(vx, vy, omega_z, current_theta, dt, p);
        
    let alpha_steer = caseIn.alpha_steer;
    if (!alpha_steer || alpha_steer.length !== wheel_count) {
        alpha_steer = omega_steer_out.map(w => Math.min(Math.abs(w) / dt, p.swerve_target_max_alpha));
    }
    
    const { Drive_Torque, Steer_Torque, Slip_Warning, debug: dyn_debug } = 
        Swerve_Dynamics_Core(ax, ay, alpha_steer, p);
        
    const max_drive_torque = Math.max(...Drive_Torque);
    const max_steer_torque = Math.max(...Steer_Torque);
    const max_grip_usage = Math.max(...dyn_debug.grip_usage);
    const max_tire_delta = Math.max(...dyn_debug.delta);
    const max_contact_half_len = Math.max(...dyn_debug.contact_half_len);
    
    const max_grip_wheel_idx = dyn_debug.grip_usage.indexOf(max_grip_usage);
    const max_delta_wheel_idx = dyn_debug.delta.indexOf(max_tire_delta);
    
    return {
        case: { vx, vy, omega_z, ax, ay, dt, current_theta, alpha_steer },
        p,
        limits,
        v_drive: v_drive_out,
        theta_steer: theta_steer_out,
        omega_steer: omega_steer_out,
        alpha_steer,
        drive_torque: Drive_Torque,
        steer_torque: Steer_Torque,
        slip_flag: Slip_Warning,
        kin_debug,
        dyn_debug,
        max_drive_torque,
        max_steer_torque,
        max_grip_usage,
        max_tire_delta,
        max_contact_half_len,
        max_grip_wheel_idx,
        max_delta_wheel_idx,
        wheel_names: wheel_count === 3 ? ["F", "RR", "RL"] : ["FL", "FR", "RR", "RL"]
    };
}